boring/src/aead.rs

```diff
@@ -0,0 +1,357 @@
+            ))?;
+            let aeadcrypter = AeadCrypter {
+                ctx,
+                max_overhead: aead.max_ciphertext_overhead(),
+            };
+            Ok(aeadcrypter)
+        }
+    }
+
+    /// Seals data by encrypting and authenticating the provided data and authenticating the provided aad.
+    ///
+    /// Returns the number of bytes written to output.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `output.len() > c_int::max_value()`.
+    ///
+    /// Panics if `output.len() < data.len() + aead.max_overhead()`.
+    pub fn seal(
+        &mut self,
+        nonce: &[u8],
+        aad: &[u8],
+        data: &[u8],
+        output: &mut [u8],
+    ) -> Result<usize, ErrorStack> {
+        unsafe {
+            assert!(output.len() <= c_int::max_value() as usize);
+            assert!(output.len() >= data.len() + self.max_overhead);
+
+            let mut output_len = output.len();
+            let nonce_len = nonce.len();
+            let data_len = data.len();
+            let aad_len = aad.len();
+
+            cvt(ffi::EVP_AEAD_CTX_seal(
+                self.ctx,
+                output.as_mut_ptr(),
+                &mut output_len,
+                output_len,
+                nonce.as_ptr(),
+                nonce_len,
+                data.as_ptr(),
+                data_len,
+                aad.as_ptr(),
+                aad_len,
+            ))?;
+
+            Ok(output_len as usize)
+        }
+    }
+
+    /// Opens data by authenticating the provided aad and authenticating and decrypting the provided data.
+    ///
+    /// Returns the number of bytes written to output.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `output.len() > c_int::max_value()`.
+    ///
+    /// Panics if `output.len() < data.len()`.
+    pub fn open(
+        &mut self,
+        nonce: &[u8],
+        aad: &[u8],
+        data: &[u8],
+        output: &mut [u8],
+    ) -> Result<usize, ErrorStack> {
+        unsafe {
+            assert!(output.len() <= c_int::max_value() as usize);
+            assert!(output.len() >= data.len());
+
+            let mut output_len = output.len();
+            let nonce_len = nonce.len();
+            let data_len = data.len();
+            let aad_len = aad.len();
+
+            cvt(ffi::EVP_AEAD_CTX_open(
+                self.ctx,
+                output.as_mut_ptr(),
+                &mut output_len,
+                output_len,
+                nonce.as_ptr(),
+                nonce_len,
+                data.as_ptr(),
+                data_len,
+                aad.as_ptr(),
+                aad_len,
+            ))?;
+
+            Ok(output_len)
+        }
+    }
+}
+
+impl Drop for AeadCrypter {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::EVP_AEAD_CTX_free(self.ctx);
+        }
+    }
+}
+
+/// Encrypts the provided data and authenticates the ciphertext and the provided aad.
+///
+/// Returns the sealed data as a vector.
+pub fn encrypt_aead(
```

Comment:
should this be an inherent method on `Aead` maybe?

Reply:
I agree. I defined them separately for parity with `symm.rs`.

https://github.com/cloudflare/boring/blob/e6ddc403337de031e446607cd161bcc41f2610db/boring/src/symm.rs#L635-L685